### ISSUE TITLE
Save As Widget exception in createWidget (async issue?)

### ISSUE THREAD

**Issue by JonFerraiolo** [maqetta/maqetta#3533]
@childsb 

I set this to Priority High to investigate, but I'm not certain this is a ship-stopper.
- New Desktop App
- Add a DIV, and inside the DIV, add an HTML button
- Select the DIV
- Save As group=a, widget=r

Result=> Dies in parser.js. Here is the full stack trace.

```
(anonymous function) (parser.js:72)
_578.map (dojo.js:15)
_1a._instantiate (parser.js:69)
_1a.instantiate (parser.js:67)
declare.constructor (DijitWidget.js:50)
(anonymous function) (declare.js:384)
widgetObject.createWidget (widget.js:551)
declare.run.dojo.forEach.position (ReplaceAction.js:37)
ret.withDoc (window.js:121)
declare.run (ReplaceAction.js:36)
array.forEach (array.js:244)
declare.run (ReplaceAction.js:30)
dt.createDijiFromNewDialog (resource.js:42)
(anonymous function) (Dialog.js:168)
lang.hitch (lang.js:375)
lang.hitch (lang.js:375)
target.(anonymous function).dispatcher (aspect.js:94)
declare._onSubmit (_DialogMixin.js:54)
declare._onClick (_ButtonMixin.js:48)
inherited (declare.js:189)
declare._onClick (Button.js:64)
lang.hitch (lang.js:373)
handles (a11yclick.js:84)

```

If you subsequently try to add the newly created widget (q.r), it works just fine, and when I set a breakpoint in createWidget() in widget.js, it looks like the same data is passed in both when createWidget is called by ReplaceAction as when it is called after dragging/dropping from widget palette.

One thing that looks very suspicious is the network panel. At the time the exception occurs, the PUT requests for the various files for the custom widget are all still pending. Could it be that the PUT requests are async, and as a result the ReplaceAction calls createWidget before the server has processed the PUT requests?

**Comment by peller** [comment]
I'm unable to get this to work, even on the second try.  What were the PUT requests which were hanging?  Folder.createResource was never converted to async IIRC.  Any idea where the global `a.r` reference should be created? (the exception is because it never gets defined)

**Comment by JonFerraiolo** [comment]
> What were the PUT requests which were hanging?

The PUT requests for foo_oam.jon, foo.js and foo.html - they weren't completed in time for the createWidget() call from ReplaceAction.js.

> Any idea where the global a.r reference should be created?

You mean where is modulePath set up for `a`? I don't know, but I would guess it's done in ReplaceAction.js somewhere. @childsb, do you remember?

**Comment by peller** [comment]
Ok, I'm definitely not seeing delays in those put requests, but I will examine the code to see how a timing problem might exist there.

As for a.r, I meant to ask where the global JS object "a" and property "r" were created.

**Comment by JonFerraiolo** [comment]
In the save as widget dialog, "a" is the widget group and "r" is the widget name. (Is that what you were asking?)

**Comment by peller** [comment]
appears to have been fixed by 9af65498e7cf6547a30cd87e692c6004eaf06985
